odoo/orm/domains.py

```diff
@@ -0,0 +1,1585 @@
+        """
+        result: list[Domain] = []
+        for c in children:
+            if isinstance(c, cls):
+                # same class, flatten
+                result.extend(c.children)
+            elif isinstance(c, DomainBool):
+                if c.value == cls.ZERO.value:
+                    continue
+                return ~cls.ZERO
+            else:
+                result.append(c)
+        len_result = len(result)
+        if len_result == 0:
+            return cls.ZERO
+        if len_result == 1:
+            return result[0]
+        return result
+
+    def __iter__(self):
+        yield from itertools.repeat(self.OPERATOR, len(self.children) - 1)
+        for c in self.children:
+            yield from c
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, DomainNary)
+            and self.OPERATOR == other.OPERATOR
+            and self.children == other.children
+        )
+
+    def __hash__(self):
+        return hash(self.OPERATOR) ^ hash(tuple(self.children))
+
+    @classmethod
+    def inverse_nary(cls):
+        """Return the inverted nary type, AND/OR"""
+        raise NotImplementedError
+
+    def __invert__(self):
+        cls = self.inverse_nary()
+        return cls([~c for c in self.children])
+
+    def iter_conditions(self):
+        for child in self.children:
+            yield from child.iter_conditions()
+
+    def map_conditions(self, function) -> Domain:
+        updated = False
+        new_children: list[Domain] = []
+        for child in self.children:
+            new_child = child.map_conditions(function)
+            if new_child is child:
+                new_children.append(child)
+            else:
+                new_children.append(new_child)
+                updated = True
+        if updated:
+            return self.new(new_children)
+        return self
+
+    def _optimize(self, model: BaseModel) -> Domain:
+        """Optimization step.
+
+        Optimize all children with the given model.
+        Run the registered optimizations until a fixed point is found.
+        See :function:`nary_optimization` for details.
+        """
+        # check if already optimized
+        if self._model_optimized:
+            if model._name == self._model_optimized:
+                return self
+            _logger.warning(
+                "Optimizing with different models %s and %s",
+                self._model_optimized, model._name,
+            )
+        assert isinstance(self, DomainNary)
+        cls = type(self)  # cls used for optimizations and rebuilding
+        # optimize children
+        children: Iterator[Domain] | Iterable[Domain] | Domain = (c._optimize(model) for c in self.children)
+        while True:
+            children = self._simplify_children(children)
+            if isinstance(children, Domain):
+                return children
+            children.sort(key=_optimize_nary_sort_key)
```

Comment:
I am still not convinced about this way to do it. The order key higly depends of the registered optimizations and then breaks the abstraction of your registration decorator.

Reply:
It really simplifies a lot of code. We sort and only create an iterator per optimization, not sure how you can group efficiently otherwise without the code becoming ugly while avoiding instantiating a dict with lists per field (and even that is not enough).
It is documented and operators are grouped together by the sorting - if anything fancier is needed, you can call the generic registration to receive all the children of the domain.

We can check this and alternatives next week.